Adding support for memory swap settings for services

wk8 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Adds support for memory swap options in services. That includes the API
plumbing, reflected in the API's doc. Added unit tests on the new feature.

**- How I did it**
Adds 2 new fields to service container specs, equivalent to `docker run`'s
`--memory-swap` and `--memory-swappiness`.

**- How to verify it**
Includes integration tests.

**- Related issues**
https://github.com/moby/moby/issues/34654

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Adds support for memory swap options in services.

cpuguy83 commented:
Let's be careful about how we expose these settings.
These are very low level and platform specific, and often confusing to users.

GordonTheTurtle commented:
<!-- AUTOMATED:POULE:DCO-EXPLANATION -->
Please sign your commits following these rules:
https://github.com/moby/moby/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "wk8/memory_flags_for_swarm" git@github.com:wk8/moby.git somewhere
$ cd somewhere
$ git rebase -i HEAD~842354426360
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Amending updates the existing PR. You **DO NOT** need to open a new one.

wk8 commented:
Please don't review yet, debugging why the tests fail on Jenkins but not in my dev container

wk8 commented on integration/service/create_test.go:
```diff
@@ -309,6 +282,90 @@ func TestCreateServiceConfigFileMode(t *testing.T) {
+	// verify that the container has the swap option set
+	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+	assert.NilError(t, err)
+	//assert.DeepEqual(t, ctnr.HostConfig.MemorySwap, swap)
```
@anshulpundir that's the failing test. No timing issue at all :( it's set to -1 when failing...

thaJeztah commented:
@wk8 I see you did a "merge" to bring changes from master in; could you do a _rebase_ instead? Let me know if you need a hand on that.

I added this to the list for tomorrow's maintainers meeting to discuss this feature; I was told at some point that memoryswap was not exposed for services because;

- swap is generally not used in production (hence, out of scope for swarm services)
- if implemented for swarm, it was desirable to have a more "generic" (and possibly platform-independent) way to configure this (which was also the reason for implementing the `--limit-cpu` and `--limit-memory` options, instead of exposing the "low-level" options.

wk8 commented:
@thaJeztah rebased as asked :)

David Yu, PM for Swarm, said that there were requests for this feature. Also, there exist a number of issues asking for this: https://github.com/docker/orca/issues/9651, https://github.com/docker/escalation/issues/726, https://github.com/docker/orca/issues/14087, https://github.com/moby/moby/issues/34654.

As to not exposing the low-level options, that's an interesting point. Happy to chat more about this!

thaJeztah commented on api/types/swarm/container.go:
```diff
@@ -66,10 +66,12 @@ type ContainerSpec struct {
+	Configs          []*ConfigReference  `json:",omitempty"`
+	Isolation        container.Isolation `json:",omitempty"`
+	Sysctls          map[string]string   `json:",omitempty"`
+	MemorySwap       int64               `json:",omitempty"`
```
Does this make the default `0` for swap? I think the current default will be "unlimited", or is it `0` already? (didn't try yet)

thaJeztah replied:
Perhaps we can use a pointer here, so that `nil` is "unset" (equal to `-1` for the container API)??

wk8 replied:
Given how it's then fed to the daemon's executor in daemon/cluster/executor/container/container.go (https://github.com/moby/moby/pull/37872/files#diff-9651468cafd3eee09570f3733efd7dc4), prior to this patch we just create an empty `enginecontainer.Resources{}` struct (https://github.com/moby/moby/blob/v17.03.2-ce/daemon/cluster/executor/container/container.go#L400), so its `MemorySwap` field is 0, and we never change it.

So the current behaviour is already defaulting to 0.

thaJeztah commented:
There's one issue with memory limits / swap; by default, it's not enabled on most hosts. In that case, the configuration will not be effective (but the user may not know). I wonder how much work it would be to take support for this option into account in the scheduler (i.e., don't schedule on a node that doesn't support setting swap).

wk8 commented:
@thaJeztah : thanks for the review!

Replied to your comment on the code.

As to

> I wonder how much work it would be to take support for this option into account in the scheduler (i.e., don't schedule on a node that doesn't support setting swap).

I agree that'd be a very-nice-to-have feature, but I'm afraid we don't have any support right now for swarmkit's agent automatically assessing an host's capabilities and reporting them back to managers (@anshulpundir or @dperny please correct me if I'm wrong!); so that'd be a much larger change. Might be one we do want to do at some point, but in my mind quite outside the scope of this patch.

thaJeztah commented:
> I agree that'd be a very-nice-to-have feature, but I'm afraid we don't have any support right now for swarmkit's agents automatically assessing their hosts' capabilities and reporting them back to managers

The `/info` endpoint does return information about this; https://docs.docker.com/engine/api/v1.37/#operation/SystemInfo - could that information be used to add additional information to nodes in a swarm?

wk8 commented:
@thaJeztah thanks for the pointer. However, I don't think Swarm currently uses any of that info for scheduling purposes. I very much agree that it's a very good idea to do just that, but also really think that this is rather outside of the scope of this change, which is simply about allowing users to set swap parameters for now. Using host info for scheduling purposes would be its own change, that would require a decent amount of further thinking about its design.

wk8 commented:
As discussed with @thaJeztah , will do the rebase when the review looks otherwise good, as I've been doing quite a few rebases for this one already :)

wk8 commented:
@thaJeztah could you please bring this up at the next maintainers' meeting? Thanks

kolyshkin commented on api/swagger.yaml:
```diff
@@ -2764,6 +2764,16 @@ definitions:
             additionalProperties:
               type: "string"
+          MemorySwap:
+            description: "Total memory limit (memory + swap). Set as `-1` to enable unlimited swap."
```
`s/Set as/Set to/`
(I know there's another similar sentence in there, but "Set to" prevails)

kolyshkin commented on api/swagger.yaml:
```diff
@@ -2764,6 +2764,16 @@ definitions:
+            type: "integer"
+            format: "int64"
+          MemorySwappiness:
+            description: "Tune a container's memory swappiness behavior. Accepts an integer between 0 and 100."
```
Would be good to describe which way it goes (i.e. if larger values means being more or less "swappy").

wk8 commented:
@thaJeztah : amended as previously discussed.

The tests I added should cover all the scenarii; but for good conscience I also manually tested all the cases we talked about using this PR's branch as well as https://github.com/docker/cli/compare/master...wk8:wk8/service_swap_options for the CLI.

* with a memory limit and no swap option:
```
# for docker run:
$ docker run --rm -m $(( 40 * 1024 * 1024 )) nginx grep hierarchical_mem /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 41943040 # 40 Mb
hierarchical_memsw_limit 83886080 # 2 * 40 Mb

# and for docker service create:
$ docker service create --limit-memory $(( 40 * 1024 * 1024 )) --name $SERVICE_NAME nginx
$ CONTAINER_ID=$(docker inspect $(docker service ps -q $SERVICE_NAME) -f '{{ .Status.ContainerStatus.ContainerID }}')
$ docker exec $CONTAINER_ID grep hierarchical_mem /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 41943040 # 40 Mb
hierarchical_memsw_limit 83886080 # 2 * 40 Mb
```

* with both a memory limit and a swap option:
```
# for docker run:
$ docker run --rm -m $(( 40 * 1024 * 1024 )) --memory-swap $(( 60 * 1024 * 1024 )) nginx grep hierarchical_mem /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 41943040 # 40 Mb
hierarchical_memsw_limit 62914560 # 40 + 20 Mb

# and for docker service create:
$ docker service create --limit-memory $(( 40 * 1024 * 1024 )) --swap-bytes $(( 20 * 1024 * 1024 )) --name $SERVICE_NAME nginx
$ CONTAINER_ID=$(docker inspect $(docker service ps -q $SERVICE_NAME) -f '{{ .Status.ContainerStatus.ContainerID }}')
$ docker exec $CONTAINER_ID grep hierarchical_mem /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 41943040 # 40 Mb
hierarchical_memsw_limit 62914560 # 40 + 20 Mb40 Mb
```

* with a memory limit, and 0 swap:
```
# for docker run:
$ docker run --rm -m $(( 40 * 1024 * 1024 )) --memory-swap $(( 40 * 1024 * 1024 )) nginx grep hierarchical_mem /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 41943040 # 40 Mb
hierarchical_memsw_limit 41943040 # 40 Mb

# and for docker service create:
$ docker service create --limit-memory $(( 40 * 1024 * 1024 )) --swap-bytes 0 --name $SERVICE_NAME nginx
$ CONTAINER_ID=$(docker inspect $(docker service ps -q $SERVICE_NAME) -f '{{ .Status.ContainerStatus.ContainerID }}')
$ docker exec $CONTAINER_ID grep hierarchical_mem /sys/fs/cgroup/memory/memory.stat
hierarchical_memory_limit 41943040 # 40 Mb
hierarchical_memsw_limit 41943040 # 40 Mb
```

wk8 commented:
@thaJeztah : rebased - could you please have another look? Thanks!

thaJeztah commented on integration/service/create_test.go:
```diff
@@ -447,3 +412,237 @@ func TestCreateServiceSysctls(t *testing.T) {
+	}
+}
+
+func TestCreateServiceWithUnlimitedMemorySwap(t *testing.T) {
```
Something like this perhaps; (after renaming `createServiceWithSingleInstanceAndWaitConverge ` to `createServiceAndConverge`)

```go
func TestCreateServiceMemorySwap(t *testing.T) {
	skip.If(
		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
		"setting service swap is unsupported before api v1.40",
	)
	skip.If(t, !testEnv.DaemonInfo.SwapLimit)

	defer setupTest(t)()
	d := swarm.NewSwarm(t, testEnv)
	defer d.Stop(t)
	client := d.NewClientT(t)
	defer client.Close()
	ctx := context.Background()

	toPtr := func(v int64) *int64 { return &v }

	tests := []struct {
		doc         string
		memorySwap  *int64
		memoryLimit int64
		expected    int64
		expectedCg  string
		expectedErr string
	}{
		{
			doc: "default",
		},
		{
			doc:         "memory-limit and memory-swap",
			memorySwap:  toPtr(1 * units.MiB),
			memoryLimit: 20 * units.MiB,
			expected:    21 * units.MiB,
			expectedCg:  fmt.Sprintf("%d", 21*units.MiB),
		},
		{
			doc:         "memory-limit",
			memoryLimit: 20 * units.MiB,
			expected:    40 * units.MiB,
			expectedCg:  fmt.Sprintf("%d", 40*units.MiB),
		},
		{
			doc:         "memory-limit and zero memory-swap",
			memorySwap:  toPtr(0),
			memoryLimit: 20 * units.MiB,
			expected:    20 * units.MiB,
			expectedCg:  fmt.Sprintf("%d", 20*units.MiB),
		},
		{
			doc:         "memory-limit and unlimited memory-swap",
			memorySwap:  toPtr(-1),
			memoryLimit: 20 * units.MiB,
			expected:    -1,
			expectedCg:  "", // maximum value depends on pagesize and host configuration
		},
		{
			doc:         "memory-swap only",
			memorySwap:  toPtr(10 * units.MiB),
			expectedErr: "Please set the memory limit when setting the swap size, see usage",
		},
	}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.doc, func(t *testing.T) {
			if tc.expectedErr != "" {
				_, err := client.ServiceCreate(
					context.Background(),
					swarm.CreateServiceSpec(t, swarm.ServiceWithMemorySwap(tc.memorySwap, tc.memoryLimit)),
					types.ServiceCreateOptions{},
				)
				assert.ErrorContains(t, err, tc.expectedErr)
				return
			}

			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwap(tc.memorySwap, tc.memoryLimit))
			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
			assert.NilError(t, err)

			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
			assert.NilError(t, err)

			if tc.memorySwap == nil {
				assert.Check(t, is.Nil(task.Spec.Resources.SwapBytes))
				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.SwapBytes))
			} else {
				assert.Check(t, is.Equal(*task.Spec.Resources.SwapBytes, *tc.memorySwap))
				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.SwapBytes, *tc.memorySwap))
			}

			assert.Check(t, is.Equal(ctnr.HostConfig.Resources.MemorySwap, tc.expected))

			if tc.expectedCg != "" {
				res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes"})
				assert.NilError(t, err)
				assert.Assert(t, is.Len(res.Stderr(), 0))
				assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)
			}

		})
	}
}
```

wk8 replied:
👍

thaJeztah commented on integration/service/create_test.go:
```diff
@@ -447,3 +412,237 @@ func TestCreateServiceSysctls(t *testing.T) {
+}
+
+// Tests that if the swappiness is not set, it should be reported as nil
+func TestCreateServiceWithNoMemorySwappiness(t *testing.T) {
```
Could be something like this;

I noticed that we may need some testing for out-of-range values (`swappiness  -1`, `swappiness 200`)


```go
func TestCreateServiceMemorySwappiness(t *testing.T) {
	skip.If(
		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
		"setting service swap is unsupported before api v1.40",
	)
	skip.If(t, !testEnv.DaemonInfo.SwapLimit)

	defer setupTest(t)()
	d := swarm.NewSwarm(t, testEnv)
	defer d.Stop(t)
	client := d.NewClientT(t)
	defer client.Close()
	ctx := context.Background()

	toPtr := func(v int64) *int64 { return &v }

	tests := []struct {
		doc        string
		swappiness *int64
		expected   int64
		expectedCg string
	}{
		{doc: "default", expectedCg: "60"},
		{doc: "zero memory-swappiness", swappiness: toPtr(0), expectedCg: "0"},
		{doc: "memory-swappiness", swappiness: toPtr(88), expectedCg: "88"},
	}

	for _, tc := range tests {
		tc := tc
		t.Run(tc.doc, func(t *testing.T) {

			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwappiness(tc.swappiness))
			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
			assert.NilError(t, err)

			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
			assert.NilError(t, err)

			if tc.swappiness == nil {
				assert.Check(t, is.Nil(ctnr.HostConfig.Resources.MemorySwappiness))
				assert.Check(t, is.Nil(task.Spec.Resources.MemorySwappiness))
				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.MemorySwappiness))
			} else {
				assert.Check(t, is.Equal(*ctnr.HostConfig.Resources.MemorySwappiness, *tc.swappiness))
				assert.Check(t, is.Equal(*task.Spec.Resources.MemorySwappiness, *tc.swappiness))
				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.MemorySwappiness, *tc.swappiness))
			}

			res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.swappiness"})
			assert.NilError(t, err)
			assert.Assert(t, is.Len(res.Stderr(), 0))
			assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)

		})
	}
}
```

thaJeztah replied:
Full diff

<details>

```patch
diff --git a/integration/internal/swarm/service.go b/integration/internal/swarm/service.go
index 365f345b90..8d95e696b8 100644
--- a/integration/internal/swarm/service.go
+++ b/integration/internal/swarm/service.go
@@ -176,17 +176,17 @@ func ServiceWithSysctls(sysctls map[string]string) ServiceSpecOpt {
 
 // ServiceWithMemorySwap sets the memory swap value of the service's TaskSpec
 // it also requires setting a memory limit, since the engine enforces that
-func ServiceWithMemorySwap(swap, memoryLimit int64) ServiceSpecOpt {
+func ServiceWithMemorySwap(swap *int64, memoryLimit int64) ServiceSpecOpt {
 	return withEnsureResourceLimits(func(spec *swarmtypes.ServiceSpec) {
-		spec.TaskTemplate.Resources.SwapBytes = &swap
+		spec.TaskTemplate.Resources.SwapBytes = swap
 		spec.TaskTemplate.Resources.Limits.MemoryBytes = memoryLimit
 	})
 }
 
 // ServiceWithMemorySwappiness sets the memory swappiness value of the service's ContainerSpec
-func ServiceWithMemorySwappiness(swappiness int64) ServiceSpecOpt {
+func ServiceWithMemorySwappiness(swappiness *int64) ServiceSpecOpt {
 	return withEnsureResourceLimits(func(spec *swarmtypes.ServiceSpec) {
-		spec.TaskTemplate.Resources.MemorySwappiness = &swappiness
+		spec.TaskTemplate.Resources.MemorySwappiness = swappiness
 	})
 }
 
diff --git a/integration/service/create_test.go b/integration/service/create_test.go
index 8b51465f5f..fd98444c53 100644
--- a/integration/service/create_test.go
+++ b/integration/service/create_test.go
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"net/http"
+	"strings"
 	"testing"
 	"time"
 
@@ -13,11 +14,11 @@ import (
 	swarmtypes "github.com/docker/docker/api/types/swarm"
 	"github.com/docker/docker/api/types/versions"
 	"github.com/docker/docker/client"
+	"github.com/docker/docker/integration/internal/container"
 	"github.com/docker/docker/integration/internal/network"
 	"github.com/docker/docker/integration/internal/swarm"
 	"github.com/docker/docker/internal/test/daemon"
 	"github.com/docker/docker/internal/test/request"
-	"github.com/docker/docker/pkg/sysinfo"
 	"github.com/docker/go-units"
 	"gotest.tools/assert"
 	is "gotest.tools/assert/cmp"
@@ -192,7 +193,7 @@ func TestCreateWithDuplicateNetworkNames(t *testing.T) {
 
 	// Create Service with the same name
 	serviceName := "top_" + t.Name()
-	serviceID, _ := createServiceWithSingleInstanceAndWaitConverge(context.Background(), t, d, client,
+	serviceID, _ := createServiceAndConverge(context.Background(), t, d, client,
 		swarm.ServiceWithName(serviceName),
 		swarm.ServiceWithNetwork(name),
 	)
@@ -241,7 +242,7 @@ func TestCreateServiceSecretFileMode(t *testing.T) {
 	assert.NilError(t, err)
 
 	serviceName := "TestService_" + t.Name()
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+	serviceID, task := createServiceAndConverge(ctx, t, d, client,
 		swarm.ServiceWithName(serviceName),
 		swarm.ServiceWithCommand([]string{"/bin/sh", "-c", "ls -l /etc/secret || /bin/top"}),
 		swarm.ServiceWithSecret(&swarmtypes.SecretReference{
@@ -293,7 +294,7 @@ func TestCreateServiceConfigFileMode(t *testing.T) {
 	assert.NilError(t, err)
 
 	serviceName := "TestService_" + t.Name()
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+	serviceID, task := createServiceAndConverge(ctx, t, d, client,
 		swarm.ServiceWithName(serviceName),
 		swarm.ServiceWithCommand([]string{"/bin/sh", "-c", "ls -l /etc/config || /bin/top"}),
 		swarm.ServiceWithConfig(&swarmtypes.ConfigReference{
@@ -368,7 +369,7 @@ func TestCreateServiceSysctls(t *testing.T) {
 		// store the map we're going to be using everywhere.
 		expectedSysctls := map[string]string{"net.ipv4.ip_nonlocal_bind": expected}
 
-		serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
+		serviceID, task := createServiceAndConverge(ctx, t, d, client,
 			swarm.ServiceWithSysctls(expectedSysctls))
 
 		// we're going to check 3 things:
@@ -413,232 +414,174 @@ func TestCreateServiceSysctls(t *testing.T) {
 	}
 }
 
-func TestCreateServiceWithMemorySwap(t *testing.T) {
+func TestCreateServiceMemorySwap(t *testing.T) {
 	skip.If(
 		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
 		"setting service swap is unsupported before api v1.40",
 	)
+	skip.If(t, !testEnv.DaemonInfo.SwapLimit)
 
 	defer setupTest(t)()
 	d := swarm.NewSwarm(t, testEnv)
 	defer d.Stop(t)
 	client := d.NewClientT(t)
 	defer client.Close()
-
 	ctx := context.Background()
-	memoryLimit := int64(20 * units.MiB)
-	swap := int64(1 * units.MiB)
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwap(swap, memoryLimit))
-	// verify that the container has the swap option set - provided that the host
-	// supports it! see
-	// https://github.com/moby/moby/blob/v17.03.2-ce/daemon/daemon_unix.go#L290-L294
-	if sysinfo.New(true).SwapLimit {
-		ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-		assert.NilError(t, err)
-		assert.Equal(t, ctnr.HostConfig.MemorySwap, memoryLimit+swap)
-	}
-
-	// verify that the task has the swap option set in the task object
-	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
-		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
-	}
-
-	// verify that the service also has the swap set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
-	}
-}
-
-func TestCreateServiceWithMemorySwapEqualToZero(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swap is unsupported before api v1.40",
-	)
-
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
 
-	ctx := context.Background()
-	memoryLimit := int64(20 * units.MiB)
-	swap := int64(0)
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwap(swap, memoryLimit))
-	// verify that the container has the swap option set - provided that the host
-	// supports it, same as `TestCreateServiceWithMemorySwap` above
-	if sysinfo.New(true).SwapLimit {
-		ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-		assert.NilError(t, err)
-		assert.Equal(t, ctnr.HostConfig.MemorySwap, memoryLimit)
+	toPtr := func(v int64) *int64 { return &v }
+
+	tests := []struct {
+		doc         string
+		memorySwap  *int64
+		memoryLimit int64
+		expected    int64
+		expectedCg  string
+		expectedErr string
+	}{
+		{
+			doc: "default",
+		},
+		{
+			doc:         "memory-limit and memory-swap",
+			memorySwap:  toPtr(1 * units.MiB),
+			memoryLimit: 20 * units.MiB,
+			expected:    21 * units.MiB,
+			expectedCg:  fmt.Sprintf("%d", 21*units.MiB),
+		},
+		{
+			doc:         "memory-limit",
+			memoryLimit: 20 * units.MiB,
+			expected:    40 * units.MiB,
+			expectedCg:  fmt.Sprintf("%d", 40*units.MiB),
+		},
+		{
+			doc:         "memory-limit and zero memory-swap",
+			memorySwap:  toPtr(0),
+			memoryLimit: 20 * units.MiB,
+			expected:    20 * units.MiB,
+			expectedCg:  fmt.Sprintf("%d", 20*units.MiB),
+		},
+		{
+			doc:         "memory-limit and unlimited memory-swap",
+			memorySwap:  toPtr(-1),
+			memoryLimit: 20 * units.MiB,
+			expected:    -1,
+			expectedCg:  "", // maximum value depends on pagesize and host configuration
+		},
+		{
+			doc:         "memory-swap only",
+			memorySwap:  toPtr(10 * units.MiB),
+			expectedErr: "Please set the memory limit when setting the swap size, see usage",
+		},
 	}
 
-	// verify that the task has the swap option set in the task object
-	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
-		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
-	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
+			if tc.expectedErr != "" {
+				_, err := client.ServiceCreate(
+					context.Background(),
+					swarm.CreateServiceSpec(t, swarm.ServiceWithMemorySwap(tc.memorySwap, tc.memoryLimit)),
+					types.ServiceCreateOptions{},
+				)
+				assert.ErrorContains(t, err, tc.expectedErr)
+				return
+			}
+
+			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwap(tc.memorySwap, tc.memoryLimit))
+			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+			assert.NilError(t, err)
+
+			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+			assert.NilError(t, err)
+
+			if tc.memorySwap == nil {
+				assert.Check(t, is.Nil(task.Spec.Resources.SwapBytes))
+				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.SwapBytes))
+			} else {
+				assert.Check(t, is.Equal(*task.Spec.Resources.SwapBytes, *tc.memorySwap))
+				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.SwapBytes, *tc.memorySwap))
+			}
+
+			assert.Check(t, is.Equal(ctnr.HostConfig.Resources.MemorySwap, tc.expected))
+
+			if tc.expectedCg != "" {
+				res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.memsw.limit_in_bytes"})
+				assert.NilError(t, err)
+				assert.Assert(t, is.Len(res.Stderr(), 0))
+				assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)
+			}
 
-	// verify that the service also has the swap set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
+		})
 	}
 }
 
-func TestCreateServiceWithUnlimitedMemorySwap(t *testing.T) {
+func TestCreateServiceMemorySwappiness(t *testing.T) {
 	skip.If(
 		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
 		"setting service swap is unsupported before api v1.40",
 	)
+	skip.If(t, !testEnv.DaemonInfo.SwapLimit)
 
 	defer setupTest(t)()
 	d := swarm.NewSwarm(t, testEnv)
 	defer d.Stop(t)
 	client := d.NewClientT(t)
 	defer client.Close()
-
-	ctx := context.Background()
-	memoryLimit := int64(20 * units.MiB)
-	swap := int64(-1)
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwap(swap, memoryLimit))
-
-	// verify that the container has the swap option set to -1
-	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-	assert.NilError(t, err)
-	assert.Equal(t, ctnr.HostConfig.MemorySwap, swap)
-
-	// verify that the task has the swap option set in the task object
-	if assert.Check(t, task.Spec.Resources.SwapBytes != nil) {
-		assert.Equal(t, *task.Spec.Resources.SwapBytes, swap)
-	}
-
-	// verify that the service also has the swap set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.SwapBytes != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.SwapBytes, swap)
-	}
-}
-
-func TestCreateServiceWithSwapBytesAndNoMemoryLimitFails(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swappiness is unsupported before api v1.40",
-	)
-
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
-	swap := int64(10 * units.MiB)
-
-	serviceOpts := func(spec *swarmtypes.ServiceSpec) {
-		if spec.TaskTemplate.Resources == nil {
-			spec.TaskTemplate.Resources = &swarmtypes.ResourceRequirements{}
-		}
-		spec.TaskTemplate.Resources.SwapBytes = &swap
-	}
-
-	spec := swarm.CreateServiceSpec(t, serviceOpts)
-	_, err := client.ServiceCreate(context.Background(), spec, types.ServiceCreateOptions{})
-
-	assert.ErrorContains(t, err, "Please set the memory limit when setting the swap size, see usage")
-}
-
-func TestCreateServiceWithMemorySwappiness(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swappiness is unsupported before api v1.40",
-	)
-
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
-
 	ctx := context.Background()
-	swappiness := int64(88)
 
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client,
-		swarm.ServiceWithMemorySwappiness(swappiness))
-
-	// verify that the container has the swappiness option set
-	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-	assert.NilError(t, err)
-	if assert.Check(t, ctnr.HostConfig.MemorySwappiness != nil) {
-		assert.Equal(t, *ctnr.HostConfig.MemorySwappiness, swappiness)
+	toPtr := func(v int64) *int64 { return &v }
+
+	tests := []struct {
+		doc        string
+		swappiness *int64
+		expected   int64
+		expectedCg string
+	}{
+		{doc: "default", expectedCg: "60"},
+		{doc: "zero memory-swappiness", swappiness: toPtr(0), expectedCg: "0"},
+		{doc: "memory-swappiness", swappiness: toPtr(88), expectedCg: "88"},
 	}
 
-	// verify that the task has the swappiness option set in the task object
-	if assert.Check(t, task.Spec.Resources.MemorySwappiness != nil) {
-		assert.Equal(t, *task.Spec.Resources.MemorySwappiness, swappiness)
-	}
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.doc, func(t *testing.T) {
 
-	// verify that the service also has the swappiness set in the spec.
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if assert.Check(t, service.Spec.TaskTemplate.Resources.MemorySwappiness != nil) {
-		assert.Equal(t, *service.Spec.TaskTemplate.Resources.MemorySwappiness, swappiness)
-	}
-}
+			serviceID, task := createServiceAndConverge(ctx, t, d, client, swarm.ServiceWithMemorySwappiness(tc.swappiness))
+			ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+			assert.NilError(t, err)
 
-// Tests that if the swappiness is not set, it should be reported as nil
-func TestCreateServiceWithNoMemorySwappiness(t *testing.T) {
-	skip.If(
-		t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.40"),
-		"setting service swappiness is unsupported before api v1.40",
-	)
+			service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
+			assert.NilError(t, err)
 
-	defer setupTest(t)()
-	d := swarm.NewSwarm(t, testEnv)
-	defer d.Stop(t)
-	client := d.NewClientT(t)
-	defer client.Close()
-
-	ctx := context.Background()
-
-	serviceID, task := createServiceWithSingleInstanceAndWaitConverge(ctx, t, d, client)
+			if tc.swappiness == nil {
+				assert.Check(t, is.Nil(ctnr.HostConfig.Resources.MemorySwappiness))
+				assert.Check(t, is.Nil(task.Spec.Resources.MemorySwappiness))
+				assert.Check(t, is.Nil(service.Spec.TaskTemplate.Resources.MemorySwappiness))
+			} else {
+				assert.Check(t, is.Equal(*ctnr.HostConfig.Resources.MemorySwappiness, *tc.swappiness))
+				assert.Check(t, is.Equal(*task.Spec.Resources.MemorySwappiness, *tc.swappiness))
+				assert.Check(t, is.Equal(*service.Spec.TaskTemplate.Resources.MemorySwappiness, *tc.swappiness))
+			}
 
-	// shouldn't be set on the container
-	ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
-	assert.NilError(t, err)
-	assert.Check(t, ctnr.HostConfig.MemorySwappiness == nil)
+			res, err := container.Exec(ctx, client, ctnr.ID, []string{"cat", "/sys/fs/cgroup/memory/memory.swappiness"})
+			assert.NilError(t, err)
+			assert.Assert(t, is.Len(res.Stderr(), 0))
+			assert.Equal(t, strings.TrimSpace(res.Stdout()), tc.expectedCg)
 
-	// nor on the task
-	if task.Spec.Resources != nil {
-		assert.Check(t, task.Spec.Resources.MemorySwappiness == nil)
-	}
-
-	// nor on the service's spec
-	service, _, err := client.ServiceInspectWithRaw(ctx, serviceID, types.ServiceInspectOptions{})
-	assert.NilError(t, err)
-	if service.Spec.TaskTemplate.Resources != nil {
-		assert.Check(t, service.Spec.TaskTemplate.Resources.MemorySwappiness == nil)
+		})
 	}
 }
 
-func createServiceWithSingleInstanceAndWaitConverge(ctx context.Context, t *testing.T, d *daemon.Daemon, client *client.Client, opts ...swarm.ServiceSpecOpt) (serviceID string, task swarmtypes.Task) {
+func createServiceAndConverge(ctx context.Context, t *testing.T, d *daemon.Daemon, client *client.Client, opts ...swarm.ServiceSpecOpt) (serviceID string, task swarmtypes.Task) {
 	serviceID = swarm.CreateService(t, d, opts...)
 
 	// wait for the service to converge to 1 running task as expected
 	poll.WaitOn(t, swarm.RunningTasksCount(client, serviceID, 1))
 
 	// retrieve the task
-	filter := filters.NewArgs()
-	filter.Add("service", serviceID)
 	tasks, err := client.TaskList(ctx, types.TaskListOptions{
-		Filters: filter,
+		Filters: filters.NewArgs(filters.KeyValuePair{Key: "service", Value: serviceID}),
 	})
 	assert.NilError(t, err)
 	assert.Check(t, is.Equal(len(tasks), 1))
```

</details>

wk8 replied:
👍

wk8 commented on daemon/cluster/convert/service.go:
```diff
@@ -400,8 +414,21 @@ func resourcesToGRPC(res *types.ResourceRequirements) *swarmapi.ResourceRequirem
+		reqs.MemorySwappiness = int64PointerToGRPC(res.MemorySwappiness)
+
+		if reqs.SwapBytes != nil && (reqs.Limits == nil || reqs.Limits.MemoryBytes == 0) {
+			return nil, fmt.Errorf("Please set the memory limit when setting the swap size, see usage")
```
👍

wk8 commented on integration/internal/swarm/service.go:
```diff
@@ -174,10 +169,25 @@ func ServiceWithEndpoint(endpoint *swarmtypes.EndpointSpec) ServiceSpecOpt {
+
+// ServiceWithMemorySwap sets the memory swap value of the service's TaskSpec
+// it also requires setting a memory limit, since the engine enforces that
+func ServiceWithMemorySwap(swap, memoryLimit int64) ServiceSpecOpt {
```
It doesn't really make sense to use one without the other, so I'm not sure it's a good idea to keep them separate.

Changing to a pointer for swap though.

wk8 commented on integration/service/create_test.go:
```diff
@@ -447,3 +412,237 @@ func TestCreateServiceSysctls(t *testing.T) {
+	}
+}
+
+func createServiceWithSingleInstanceAndWaitConverge(ctx context.Context, t *testing.T, d *daemon.Daemon, client *client.Client, opts ...swarm.ServiceSpecOpt) (serviceID string, task swarmtypes.Task) {
```
Naming is hard ;) changed!

wk8 commented on integration/service/create_test.go:
```diff
@@ -447,3 +412,237 @@ func TestCreateServiceSysctls(t *testing.T) {
+	// verify that the container has the swap option set - provided that the host
+	// supports it! see
+	// https://github.com/moby/moby/blob/v17.03.2-ce/daemon/daemon_unix.go#L290-L294
+	if sysinfo.New(true).SwapLimit {
```
It's still worth checking that the info made it down to the service and task specs?

wk8 commented on api/types/swarm/task.go:
```diff
@@ -121,6 +121,17 @@ type DiscreteGenericResource struct {
+	// -1 means unlimited
+	// a null pointer keeps the default behaviour of granting twice the memory
+	// amount in swap
+	SwapBytes *int64
```
Except `Limits` shares the same struct as `Reservations`, and there is a case to be made to keep it that way: it seemed to make more sense to keep the symmetry there, and take fields that don't make sense as both a limit and a reservation out in the `Requirements`.

Besides, you reviewed and approved this in https://github.com/docker/swarmkit/pull/2816/files :) ?

cpuguy83 replied:
I think we could keep a base version of `Resources` for sharing between `Limits` and `Reservations` and let each diverge as needed.

Other than this, the changes in this PR look great.

wk8 commented on daemon/cluster/executor/container/container.go:
```diff
@@ -448,6 +448,20 @@ func (c *containerConfig) resources() enginecontainer.Resources {
 	if r.Limits.MemoryBytes > 0 {
 		resources.Memory = r.Limits.MemoryBytes
+
+		if r.SwapBytes != nil {
```
That would be duplicating the logic that already does this at https://github.com/moby/moby/blob/dc52692458a68ad453fc039460d1d6c7bf39b837/daemon/daemon_unix.go#L340-L343 ; seems to me it's better to have that in just one place?

wk8 commented on integration/internal/swarm/service.go:
```diff
@@ -174,10 +169,25 @@ func ServiceWithEndpoint(endpoint *swarmtypes.EndpointSpec) ServiceSpecOpt {
+}
+
+// ServiceWithMemorySwappiness sets the memory swappiness value of the service's ContainerSpec
+func ServiceWithMemorySwappiness(swappiness int64) ServiceSpecOpt {
```
👍

wk8 commented on integration/service/create_test.go:
```diff
@@ -447,3 +412,237 @@ func TestCreateServiceSysctls(t *testing.T) {
+	if sysinfo.New(true).SwapLimit {
+		ctnr, err := client.ContainerInspect(ctx, task.Status.ContainerStatus.ContainerID)
+		assert.NilError(t, err)
+		assert.Equal(t, ctnr.HostConfig.MemorySwap, memoryLimit+swap)
```
👍

wk8 commented:
@thaJeztah : rebased, and amended as requested; could you please take another look?

wk8 committed:
Adding support for memory swap settings for services

With integration tests

Relevant Swarmkit PR: https://github.com/docker/swarmkit/pull/2816
(updated the vendored version of Swarkit to that)

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@thaJeztah : could you please take another look? Thanks.

david-yu commented:
@thaJeztah Could we revisit this again? Thanks!

wk8 commented:
As per our discussion today with @thaJeztah  & @cpuguy83 :
* swap should go into the `Limits` struct
* and that in turns mean that `Limits` and `Reservations` are now going to be two separate structs
And all the changes that this entails.

wk8 committed:
Merge branch 'master' into wk8/memory_flags_for_swarm

Signed-off-by: Jean Rouge <rougej+github@gmail.com>

wk8 commented:
@thaJeztah & @cpuguy83 : before I do the changes outlined in https://github.com/moby/moby/pull/37872#issuecomment-482249014, could you please validate that https://github.com/docker/swarmkit/pull/2865/files#diff-2825f8c43cd6822a89834c265124fe46 looks good to you, and that this shape for the Limits/Resources structs should be the final one, and won't need to be changed another time? Thanks.

wk8 commented:
/retrier ignore

david-yu commented:
Is this ready to merge?

m-barthelemy commented:
Hi, is this still being worked on?

Me1kaa commented:
Any progress here?

mhaddadi commented:
Hi, just wondering if this PR is still in progress as it would be useful to be able to deactivate swapping in swarm mode ?
Thanks

everyx commented:
Any update?

D0wn3r commented:
Still not possible to deactivate swap ?

rickpeters commented:
Anybody still out here....?

futureecomavi commented:
Is anything happening with this?

demaniak commented:
Has there been any further traction on this?
